main.go

```diff
@@ -61,62 +53,169 @@ func List() {
 	Sort(entries)
 	Header()
 	for i, j := range entries {
-		fmt.Printf("%4d %14s %14s %20s\n", j.ID, j.LastName, j.FirstName, j.PhoneNumber)
+		fmt.Printf("%10d %20s %20s %20s\n", j.ID, j.LastName, j.FirstName, j.PhoneNumber)
 		if (i+1)%20 == 0 && i < len(entries) {
+			fmt.Println(" ")
 			fmt.Println("Press <ENTER> to continue...")
+			fmt.Println(" ")
 			fmt.Scanln()
 			Header()
 		}
 	}
+}
 
+func GiveID(entries []Entry) (i uint) {
+	i = uint(len(entries)) + 1 //Thats not ideal and will work wrong if the last contact will be deleted
+	return i
 }
 
-func Add(entries []Entry) {
-	// how to encode with user input???
-	f, err := os.Open("phonebook.json")
+func Add() {
+	// how to encode with user input??? Tutorial by me (version 1)
+	var entries []Entry // make variable where will be all data
+
+	f, err := os.Open("phonebook.json") // open and encode file with already been information and put it in variable you made in previous step
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer f.Close()
+	if err := json.NewDecoder(bufio.NewReader(f)).Decode(&entries); err != nil {
+		log.Fatal(err)
+	}
+
+	var newentries Entry // make new variable with type of your struct, but be careful - it need to be NOT SLICE OF STRUCT, JUST STRUCT, CAUSE WE MAKING ONLY ONE NEW CONTACT, NOT 5 IN ONE TIME
+
+	scanner := bufio.NewScanner(os.Stdin) // Scan information you need
+	fmt.Println("Enter last name:")       // better not use bufio reader, it add \n\r to text and you ll need 2 variables for string and error
+	scanner.Scan()
+	scanner.Scan()
+	newentries.LastName = scanner.Text()
+	fmt.Println("Enter first name:")
+	scanner.Scan()
+	newentries.FirstName = scanner.Text()
+	fmt.Println("Enter phone number:")
+	scanner.Scan()
+	newentries.PhoneNumber = scanner.Text()
+	newentries.ID = GiveID(entries)
+
+	entries = append(entries, newentries) // add scanned data to decoded earlier data
+
+	f, err = os.Create("phonebook.json") // and encode it back
 	if err != nil {
 		log.Fatal(err)
 	}
 	defer f.Close()
 	w := bufio.NewWriter(f)
 	defer w.Flush()
-	//reader := bufio.NewReader(os.Stdin)
 	if err := json.NewEncoder(w).Encode(entries); err != nil {
 		log.Fatal(err)
 	}
 
 }
 
-///func Remove(entries []Entry, s int) []Entry {
-// return append(s[:s], slice[s+1:]...)
-//}
+func Remove() {
+	var entries []Entry                 //var
+	f, err := os.Open("phonebook.json") // opening and decoding
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer f.Close()
```

Comment:
There's a funny thing about deferred close. The call actually happens at the end of the function! It means that the file is opened for reading until the very end (even when you already start writing to that file). It'd better to factor out reading to a separate function. It would guarantee the file is closed when you don't need it anymore. E.g.

```go
func ReadPhonebook(filename string) (entries []Entry, err error) {
  f, err := os.Open(filename)
  if err != nil {
    return nil, err
  }
  defer f.Close()
  r := bufio.NewReader(f)
  if err = json.NewDecoder(r).Decode(&entries); err != nil {
    return nil, err
  }
  return entries, nil
}
```